### PR TITLE
DMA - take two

### PR DESCRIPTION
**Pull request by thalesfragoso**
This is a rework of https://github.com/stm32-rs/stm32f4xx-hal/pull/88

I did some reorganization and also moved to @ra-kete traits that should be incorporated in some place in `rust-embedded` in the future.

This PR doesn't include any way of using it with the HAL's peripherals, it's big enough already. It should be pretty easy to add individual peripheral support in the future, just some traits that need to be implemented and maybe a helper method in the peripheral code.

However, it's already usable with almost all of the pac's peripherals, which you need to setup yourself. Sorry for the huge amount of code in one PR, but a good chunk of it is macros to implement traits for all the peripherals, which prevents using the wrong set of peripheral, memory size, stream, channel and more. The buffer interface is also safe, it always ask for a `&mut` on the buffer though, this isn't mandatory, it can be extended to accept `&` when dealing with memory to peripheral transfers, it just isn't implemented that way yet.

Circular transfers aren't implemented also, but double buffer transfer are and can be used for continuous transfers. The PR is almost ready for review and I will mark as such soon, just need to check some minor stuff.

Examples:
https://gist.github.com/thalesfragoso/a07340c5df6eee3b04c42fdc69ecdcb1
https://gist.github.com/thalesfragoso/6168e53411acb54f48dddab7b84e8576  

A good amount of the work done is the PR was made by @cs2dsb

### CODE REVIEW

**Comment by thalesfragoso** [comment]
Weird, I can build F410 locally with the last commit .-.

**Comment by therealprof** [comment]
@thalesfragoso Me, too. Weird. Maybe rebase and try again?

**Comment by thalesfragoso** [comment]
> Maybe rebase and try again?

It did the trick. Okay, now I think I just need to update some of the docs and this will be ready for review. I will try to do it today.

**Comment by therealprof** [comment]
Looks good to me on first glance. Is this ready to be taken out of Draft mode?

**Comment by thalesfragoso** [comment]
Okay, now I think that this is ready for review. @ra-kete it would be nice if you could take a look at the safety aspects of this PR, thanks.

File: src/dma/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1171 @@
...
+            assert!(
+                new_buf_len >= usize::from(self.transfer_length),
+                "Second Buffer not big enough"
+            );
+
+            // If this is false, then RAM corruption might have occurred, there's nothing we can do
+            // apart from panicking.
+            // TODO: Is this the best solution ? The closure based approach seems necessary if we
+            // want to support BBqueue.
+            assert!(current_buffer == STREAM::current_buffer(), "Overrun");
```
Comment (ra-kete) [inline src/dma/mod.rs:1116]: This check needs to be moved after the `set_memory_address` calls, I think. Only then can we be sure the DMA will not write into the buffer we just gave to the user anymore.

**Comment by cs2dsb** [comment]
Hey, just wanted to mention the use case I implemented the original DMA stuff for was using circular mode so it would be great to see that implemented. In order to hack around it in the changes you made in my fork a while back I had to expose the STREAM from the DMA transfer - I wonder if this would be useful in general? It's a mammoth task to implement every possible use case but if you do lots of the fiddly config stuff and expose the internal structs for advanced hacking behind an unsafe method that's very helpful to those of us who want to go off piste. I've felt this about other hal implementations like timers, it's useful to get it mostly configured but then having to work directly with TIMx::ptr() after the fact seems a bit ugly.

I'd be happy to discuss a nice circular implementation if that's something people are interested in. I kinda went off and hacked what in my fork since the PR lay dormant for so long I assumed no one else was interested in using DMA but if it's getting some attention I'm glad to contribute.

File: src/dma/traits.rs [hunk 2]
```diff
@@ -0,0 +1,1394 @@
...
+    fn set_priority(&mut self, priority: config::Priority);
+
+    /// Set the memory size (msize) for the DMA stream.
+    ///
+    /// # Safety
+    /// This must have the same alignment of the buffer used in the transfer.
+    /// Valid values:
+    ///     * 0 -> byte
+    ///     * 1 -> half word
+    ///     * 2 -> word
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/traits.rs:85]: Those are the bits used to write to the register, and the only valid values.

File: src/dma/traits.rs [hunk 3]
```diff
@@ -0,0 +1,1394 @@
...
+}
+
+// Convenience macro for implementing addresses on peripherals
+macro_rules! address {
+    ($(($peripheral:ty, $register:ident, $size: ty)),+ $(,)*) => {
+        $(
+            impl Sealed for $peripheral {}
+
+            impl PeriAddress for $peripheral {
+                #[inline]
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/traits.rs:178]: Hmm, not sure about the impact, first I thought of `#[inline(always)]` which I think would cause inlining in debug builds, but I thought maybe it would be too harsh ? not sure.

File: src/dma/traits.rs [hunk 4]
```diff
@@ -0,0 +1,1394 @@
...
+                type Target = T;
+
+                #[inline]
+                fn deref(&self) -> &T {
+                    &self.0
+                }
+            }
+        )+
+    };
+}
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/traits.rs:278]: The type is used, but I think you're right about the `$register` field not doing anything. I didn't actually write that code, so I left it there, but maybe we should remove the `$register` part.

File: src/dma/traits.rs [hunk 5]
```diff
@@ -0,0 +1,1394 @@
...
+    feature = "stm32f429",
+    feature = "stm32f446",
+    feature = "stm32f469",
+    feature = "stm32f479",
+))]
+dma_map!(
+    (Stream0<DMA1>, Channel4, pac::UART5, PeripheralToMemory), //UART5_RX
+    (Stream2<DMA1>, Channel4, pac::UART4, PeripheralToMemory), //UART4_RX
+    (Stream4<DMA1>, Channel4, pac::UART4, MemoryToPeripheral), //UART4_TX
+                                                               //(pac::DMA1, Stream6, Channel7, pac::DAC2, MemoryToPeripheral), //DAC2
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/traits.rs:792]: I looked at some of those, it seems that it needs more work in the peripheral itself (new types, etc) to be able to conform with our API here, I left some `TODO`s around, I remember putting a TODO for the DAC somewhere, hmmm.

File: src/dma/mod.rs [hunk 6]
```diff
@@ -0,0 +1,1171 @@
...
+                    //NOTE(unsafe) Atomic write with no side-effects and we only access the bits
+                    // that belongs to the StreamX
+                    let dma = unsafe { &*I::ptr() };
+                    dma.$ifcr.write(|w| w
+                        .$tcif().set_bit() //Clear transfer complete interrupt flag
+                        .$htif().set_bit() //Clear half transfer interrupt flag
+                        .$teif().set_bit() //Clear transfer error interrupt flag
+                        .$dmeif().set_bit() //Clear direct mode error interrupt flag
+                        .$feif().set_bit() //Clear fifo error interrupt flag
+                    );
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:287]: hmm, not sure if I agree, it isn't an unsafe operation and it would almost be warning territory.

File: src/dma/mod.rs [hunk 7]
```diff
@@ -0,0 +1,1171 @@
...
+                    dma.st[Self::NUMBER].cr.modify(|_, w| w.en().clear_bit());
+                }
+
+                #[inline]
+                fn set_channel<C: Channel>(&mut self, channel: C) {
+                    //NOTE(unsafe) We only access the registers that belongs to the StreamX
+                    let dma = unsafe { &*I::ptr() };
+                    //Some device crates have this field unsafe, others don't.
+                    #[allow(unused_unsafe)]
+                    dma.st[Self::NUMBER].cr.modify(|_, w| unsafe { w.chsel().bits(channel.bits()) });
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:399]: I think manual intervention is required on the svd to mark all valid values as safe, so not all PACs have it.

File: src/dma/mod.rs [hunk 8]
```diff
@@ -0,0 +1,1171 @@
...
+    BUF: WriteBuffer<Word = <PERIPHERAL as PeriAddress>::MemSize> + 'static,
+    (STREAM, CHANNEL, PERIPHERAL, DIR): Sealed,
+{
+    /// Applies all fields in DmaConfig.
+    fn apply_config(&mut self, config: config::DmaConfig) {
+        let msize = mem::align_of::<<PERIPHERAL as PeriAddress>::MemSize>() / 2;
+
+        self.stream.clear_interrupts();
+        self.stream.set_priority(config.priority);
+        // NOTE(unsafe) These values are correct because of the invariants of PeriAddress
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:832]: Only the HAL can implement `PeriAddress` because of the `Sealed` bound, but maybe we should mark as `unsafe` to be more concise and maybe help other people extending this API.

File: src/dma/mod.rs [hunk 9]
```diff
@@ -0,0 +1,1171 @@
...
+    /// Configures the DMA stream to the correct channel for the peripheral, configures source and
+    /// destination and applies supplied configuration. In a memory to memory transfer,
+    /// the `double_buf` argument is the source of the data.
+    pub fn init(
+        mut stream: STREAM,
+        peripheral: PERIPHERAL,
+        mut memory: BUF,
+        mut double_buf: Option<BUF>,
+        config: config::DmaConfig,
+    ) -> Result<Self, DMAError<BUF>> {
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:866]: Hmm, not sure how people feel about panics, but I think you might be right on this one.

File: src/dma/mod.rs [hunk 10]
```diff
@@ -0,0 +1,1171 @@
...
+            if new_buf_len < usize::from(self.transfer_length) {
+                return Err(DMAError::SmallBuffer(new_buf));
+            }
+
+            if STREAM::current_buffer() == CurrentBuffer::DoubleBuffer {
+                self.stream.set_memory_address(new_buf_ptr as u32);
+                let old_buf = self.buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:980]: `self.buf.replace` does a read, but you are right, the replace and all buffer operations could still be moved to before the address setting, will fix this.

File: src/dma/mod.rs [hunk 11]
```diff
@@ -0,0 +1,1171 @@
...
+                self.stream
+                    .set_memory_double_buffer_address(new_buf_ptr as u32);
+                let old_buf = self.double_buf.replace(new_buf);
+
+                // "Subsequent reads and writes cannot be moved ahead of preceding reads"
+                compiler_fence(Ordering::Acquire);
+
+                // double buffering, unwrap can never fail
+                return Ok((old_buf.unwrap(), CurrentBuffer::DoubleBuffer));
+            }
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:994]: In that case the `stream.set_memory_address(...)` would fail, but that is still a problem, since the user will be given the buffer to where the DMA is writing, the check is right before changing the memory in this case, so like a few clock cycles, this seems to be almost impossible to cause problems, also checking again after can cause some false positives, and in the case of a false positive we would have the same problem, since we would give the `new_buf` back with the error, but that buffer would be the one the DMA would be writing to, so not sure what we can do here... panic ?

File: src/dma/mod.rs [hunk 12]
```diff
@@ -0,0 +1,1171 @@
...
+        self.stream.clear_interrupts();
+
+        unsafe {
+            let stream = ptr::read(&self.stream);
+            let peripheral = ptr::read(&self.peripheral);
+            let buf = ptr::read(&self.buf);
+            let double_buf = ptr::read(&self.double_buf);
+            mem::forget(self);
+            (stream, peripheral, buf.unwrap(), double_buf)
+        }
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:1031]: > Is that done because we cannot allow Drop to run because it would access the already moved out stream?

Yes, the compiler doesn't let you destruct a type that implements `Drop`. It would give a compiler error.

> The Option approach used in the Embedonomicon might be more inconvenient to implement, but it doesn't have this drawback.

The `Option` approach has the `unwrap` operation every time you want to access the inner transfer, which is a considerable drawback IMO (performance and convenience wise), and the `ptr::read` seems to be way more likely to be optimized by the compiler, especially given that both `peripheral` and `stream` are zero sized types and the buffer is a pointer type which probably is quite small.

File: src/dma/mod.rs [hunk 13]
```diff
@@ -0,0 +1,1171 @@
...
+            };
+            let r = f(db, !current_buffer);
+            let mut new_buf = r.0;
+            let (new_buf_ptr, new_buf_len) = new_buf.write_buffer();
+
+            // We can't change the transfer length while double buffering
+            assert!(
+                new_buf_len >= usize::from(self.transfer_length),
+                "Second Buffer not big enough"
+            );
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:1110]: In this case we already gave the old buffer to the user through the closure, so we can't take it back, and we don't have anything to put in its place, so panic seems okay, we could also shutdown everything and return an error, but this seems more like a code bug instead of runtime error type thing, which would be quite inconvenient for the user to recover anyways.

File: src/dma/mod.rs [hunk 14]
```diff
@@ -0,0 +1,1171 @@
...
+        self.buf.replace(new_buf);
+
+        // "Preceding reads and writes cannot be moved past subsequent writes"
+        compiler_fence(Ordering::Release);
+
+        self.stream.enable();
+
+        Ok(r.1)
+    }
+}
```
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:1159]: Both `get_transfer_complete_flag()` and `get_half_transfer_flag()` on the stream's trait are free functions, the user can call them anytime anywhere.

**Comment by thalesfragoso** [comment]
> Hey, just wanted to mention the use case I implemented the original DMA stuff for was using circular mode so it would be great to see that implemented. 

A safe api for the buffers in circular transfers would look a bit different, I think it's even better to create a new type instead of using the common `Transfer` one, I left it to be extended in the future since this PR is quite big already.

> In order to hack around it in the changes you made in my fork a while back I had to expose the STREAM from the DMA transfer - I wonder if this would be useful in general? It's a mammoth task to implement every possible use case but if you do lots of the fiddly config stuff and expose the internal structs for advanced hacking behind an unsafe method that's very helpful to those of us who want to go off piste.

Hmm, an unsafe method to get to the stream seems plausible. 

Just for curiosity, what case would the circular transfer cover that a double buffering approach wouldn't be enough ? Double buffering seems to cover the same niche and it's more flexibly, I think.

**Comment by cs2dsb** [comment]
Doing a circular transfer is just about extracting the maximum performance in the case I'm using it - I'm doing motor control where you want to process one set of samples per interrupt and the range of motors you can support is limited by the maximum PWM frequency you can support so the less re-configuring needed in the ISR the better. Admittedly I've not benchmarked how much swapping the buffers and re-configuring the dma transfer each interrupt would add but none of the C implementations do it so it was a small fixed overhead I wanted to avoid.

**Comment by thalesfragoso** [comment]
>  Admittedly I've not benchmarked how much swapping the buffers and re-configuring the dma transfer each interrupt would add but none of the C implementations do it so it was a small fixed overhead I wanted to avoid.

There is no re-configuring, you just change the pointer to the buffer that the DMA isn't using, the DMA never stops, just like a circular transfer. The benefit is that you don't need it to be a continuous buffer, which allows you to use a memory pool, etc. It also allows you to get an owned buffer and don't worry about overruns while you're processing your data.

Circular transfers are probably more common because not all hardware supports double buffering DMA.

File: src/dma/traits.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/traits.rs:85]: I see, this API is more low-level than I initially figured.
You could make this clearer by taking a dedicated enum that implemented `Bits` to get these values. Also would avoid the need to check if the value is valid.

File: src/dma/traits.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/traits.rs:178]: I usually only add compiler hints only after actually testing if they improve anything (inlining can have negative effects on caching). But if there is an a convention how this is done in this crate or the embedded ecosystem in general, it's probably best to go with that. I don't mean to start a huge discussion here :)

File: src/dma/traits.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/traits.rs:278]: I would remove it, a) because it is unused code and you wonder if there is a bug/typo somewhere here and b) because it makes readers think that "which register of the contained timer to use for DMA" is encoded in the `$register`, when it's actually encoded in the `$name`.

I guess `$register` may be there as additional documentation in the `tim_channels!` invocation, so you see that, e.g., the wrapper type `CCR1` refers the the register `ccr1`. But IMO that's not *too* hard to guess from the name already ;)

File: src/dma/traits.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/traits.rs:792]: Yes, most of the commented out code already has comments saying why it's commented out (mostly just because there seem to be registers missing in the PAC?). Looks like there is just one missing here.

File: src/dma/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:287]: Not sure what you mean by warning territory? Do you think the compiler should throw a warning if one puts `unsafe` around an operation that's not marked unsafe itself?

In this case I would simply use it as a marker to know where the dangerous code ends, basically telling the reader "Don't randomly modify dma registers in this section." But the function is pretty small, so the scope where unsafe stuff could happen is pretty limited already, so your way is fine too.

File: src/dma/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:832]: Can we really use "it's only used internally" as an excuse for not marking code that's potentially unsafe? I can't remember dealing with such cases before, but I always thought of safety contracts and accessibility as orthogonal concerns.

HAL developers from the future can make mistakes too, so adding that as an additional documentation is helpful I think. Plus, as you said, it's more future-proof that way.

File: src/dma/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:866]: Also not sure. I guess they can add fmt bloat if they don't get optimized away. But a `result.unwrap()` does too and that's probably harder for the compiler to optimize because of the additional level of indirection.

Maybe there is some crate-wide policy regarding Results vs panics that can help us here?

File: src/dma/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:980]: You can't really rely on `self.buf.replace` because it doesn't do a volatile read, right? So it might be optimized away.

File: src/dma/mod.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:994]: We found a solution in Matrix: Since the MAR registers are write-protected if the respective buffer is currently used by DMA we can simply check if they contain the address of the new buffer after writing it. And if not, raise/return an Overrun error.

File: src/dma/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:1031]: `Option` has the drawback of being inconvenient, but only for HAL writers. Users of the API don't have to notice it. And I'm pretty sure you can define an internal helper method to make it easier inside the HAL too. Regarding performance, I don't think it would be too bad, but if you are concerned about that there are `unsafe` ways to tell the compiler that the option is always `Some` (outside of the `Drop` impl), so performance should stay the same.

But I agree that using an `Option` is a hassle so we should only use it if your approach has real drawbacks. What concerns me is that we leak memory here and I'm not sure that gets optimized away. If it doesn't you leave the user with no choice than to leak memory if they want to get their resources back. If you have an application that does continuous one-shot transfers, at some point all the available memory will be leaked, which would make the API unusable.

File: src/dma/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:1110]: I agree, didn't think about how to continue the transfer in this case.

File: src/dma/mod.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:1159]: Ah, I totally missed that. Would be good to add an example so the user can see how the API is supposed to be used.

File: src/dma/traits.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/traits.rs:85]: There is no check currently, the enum based would need a check when creating the enum though, since we got the size from `mem::align_of`. Not sure if I want to do that since it doesn't seem necessary and the API is already unsafe.

File: src/dma/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:980]: I don't think it would in this case since the return result depends on that, I also have the impression that the compiler fence would pass before that, since it usually states about program order. However, like I said I'm not sure if it would prevent reordering to before the register write, so I will go with `SeqCst`.

File: src/dma/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:1031]: I'm not sure what you think will leak here, the only types that we don't "move out" with `ptr::read` are `PhantomData` and an `u16` both of which don't implement (and never will) `Drop`.

File: src/dma/mod.rs [hunk 15]
```diff
@@ -1080,8 +1112,15 @@ where
     /// * The closure `f` takes too long to return and a buffer overrun occurs.
     ///
     /// # Safety
-    /// Memory corruption might occur if the previous buffer, the one passed to the closure, gets
-    /// dropped in the closure and an overrun occurs in double buffering mode.
+    ///
+    /// Memory corruption might occur in the previous buffer, the one passed to the closure, if an
+    /// overrun occurs in double buffering mode. This method will panic if it detects an overrun,
+    /// however, the overrun won't be detected if the closure takes long enough to cause the DMA to
+    /// change buffers twice since the call of this method.
```
Comment (ra-kete) [inline src/dma/mod.rs:1119]: Shouldn't we be able to know if DMA changed the buffers twice by looking at the interrupt flags? I guess we'd have to clear them before invoking the closure and I'm not sure if we can/should. But if we can, we can just check those flags afterwards and if one is set, that means an overrun occurred.

File: src/dma/traits.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/traits.rs:91]: It wouldn't be necessary, just more readable. But if you want to avoid that additional complexity that's okay with me.

File: src/dma/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ra-kete) [reply src/dma/mod.rs:1063]: I think my misconception was that when I saw `mem::forget` I though "leak". But in this case we only `forget` stack-allocated memory, which will be freed when that stack frame is popped, so there is no actual leak here.

So I think you are right and this is okay to do. I'd still very much like a comment describing why it is done that way. And also why it doesn't leak (so people like me don't get confused).

File: src/dma/mod.rs (hunk 15 above)
Comment (thalesfragoso) [reply src/dma/mod.rs:1119]: You are right, we should be able to detect that by checking the flags, we might also do that instead of reading the address register at the two methods, I will implement this. Also, we are already clearing the flag at the beginning of the method and I should document it.

File: src/dma/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (thalesfragoso) [reply src/dma/mod.rs:409]: Ok, I just checked and that comment doesn't hold true anymore, removed.

**Comment by thalesfragoso** [comment]
> There's a conflict and it's missing a CHANGELOG entry. Other than that looks good to me.

Great, I will fix those and test again on hardware tonight.

**Comment by thalesfragoso** [comment]
Done, I also tested on hardware and updated the gists examples (some bugs...).

File: src/dma/buffer.rs [hunk 16]
```diff
@@ -0,0 +1,219 @@
+//! Traits that represent a buffer that can be used with DMA.
+//! Copied from ra-kete's dma-poc repository.
```
Comment (therealprof) [inline src/dma/buffer.rs:2]: Just stumbled across this.

Can we replace this by something more substantial, either a permalink to the repo or some issue? It sounds like something that is going to change, too (e.g. by being replaced by an import of the finished traits at some point?), if so it would be good to point out that this is only temporary.

Comment (thalesfragoso) [reply src/dma/buffer.rs:2]: Yes, this is intended to be temporary. I will add a note about it.

File: src/prelude.rs [hunk 17]
```diff
@@ -9,7 +9,6 @@ pub use embedded_hal::prelude::*;
     not(any(feature = "stm32f411", feature = "stm32f412", feature = "stm32f401",))
 ))]
 pub use crate::dac::DacExt as _stm32f4xx_hal_dac_DacExt;
-pub use crate::dma::DmaExt as _stm32f4xx_hal_dma_DmaExt;
```
Comment (therealprof) [inline src/prelude.rs:12]: Is this usable without re-exporting the traits in the prelude?

**Comment by therealprof** [comment]
I'll leave it up to you whether you want people to see the sausage making or squash a little before we merge this. ;)

File: src/prelude.rs (hunk 17 above)
Comment (thalesfragoso) [reply src/prelude.rs:12]: This trait don't exist anymore, I changed to using the `fn new()` approach, the user doesn't need any trait in scope to use the API, they might need the `Stream` trait to use some helper methods, but it isn't usually required.

**Comment by thalesfragoso** [comment]
> I'll leave it up to you whether you want people to see the sausage making or squash a little before we merge this. ;)

I think we have enough code for 8 commits, heh. I will leave like this, might scary someone seeing 3k changes in a couple of commits.
